### PR TITLE
Fix security issue #46

### PR DESCRIPTION
**Pull request by nickoferrall**
PR to fix issue #4028 

This PR also:
- Fixes a bug where the `AddOrgMutation` wasn't handling errors in the `onComplete` function
- Fixes several bugs where error objects were being used in the jsx instead of a string
- Refactored a few class components to functional components 

I'll leave the testing criteria out of this public PR description as it's a security issue.

### CODE REVIEW

File: packages/server/graphql/mutations/addTeam.ts [hunk 1]
```diff
@@ -45,7 +45,10 @@ export default {
...
         .run()
 
-      const orgTeamNames = orgTeams.map((team) => !team.isArchived && team.name)
+      const orgTeamNames = [] as string[]
+      orgTeams.forEach((team) => {
+        if (!team.isArchived) orgTeamNames.push(team.name)
+      })
       const {
         data: {newTeam},
         errors
```
Comment (nickoferrall) [inline packages/server/graphql/mutations/addTeam.ts:54]: Why is `addTeamValidation` a thunk out of interest?

File: packages/client/modules/newTeam/components/NewTeamForm/NewTeamForm.tsx [hunk 2]
```diff
@@ -88,53 +86,33 @@ interface Fields {
...
-class NewTeamForm extends Component<Props, State> {
-  state = {
-    isNewOrg: this.props.isNewOrganization,
-    orgId: '',
-    fields: {
-      orgName: {...DEFAULT_FIELD},
-      teamName: {...DEFAULT_FIELD}
-    }
-  }
+const NewTeamForm = (props: Props) => {
```
Comment (mattkrick) [inline packages/client/modules/newTeam/components/NewTeamForm/NewTeamForm.tsx:89]: nice refactor!

File: packages/client/modules/newTeam/components/NewTeamForm/NewTeamForm.tsx [hunk 3]
```diff
@@ -88,53 +86,33 @@ interface Fields {
...
+  const {isNewOrganization, organizations} = props
```
Comment (mattkrick) [inline packages/client/modules/newTeam/components/NewTeamForm/NewTeamForm.tsx:90]: nit: maybe call this `isInitiallyNewOrg`? something so we don't confuse it with what we're using for state

File: packages/client/modules/newTeam/components/NewTeamForm/NewTeamForm.tsx [hunk 4]
```diff
@@ -146,96 +124,88 @@ class NewTeamForm extends Component<Props, State> {
...
-      .max(100, 'Maybe just the legal name?')
+  const validate = (name: FieldName) => {
+    const validators = {
+      teamName: validateTeamName,
+      orgName: validateOrgName
+    }
+    const res: Legitity = validators[name]()
+    const field = fields[name]
+    if (res.error !== field.error) {
+      setFields({
```
Comment (mattkrick) [inline packages/client/modules/newTeam/components/NewTeamForm/NewTeamForm.tsx:135]: nit: we've got a nice hook called `useForm` that takes care of all this boilerplate logic for you if you'd like to use that

File: packages/server/graphql/mutations/addTeam.ts (hunk 1 above)
Comment (mattkrick) [reply packages/server/graphql/mutations/addTeam.ts:54]: me from years ago was a terrible programmer

**Review by mattkrick — COMMENTED** [review]
looks really good!
only big suggestion is to try using `useForm` & see if that cleans up some of the boilerplate of using a form.

forms are so hard & it seems like everyone loves building their own form packages. thankfully, it's super rare for us to have large forms. i think this is our second largest form in the whole app!

let me know if you find `useForm` helpful, we could still make some fixes to it to solve our simple use cases. 

or if time is tight just let me know & we can push this through without. your call!

**Comment by nickoferrall** [comment]
Refactored to use `useForm` - it was very soothing removing that code!

**Comment by mattkrick** [comment]
linkify is a great tool because it very aggressively makes things links.
for example,  `foo.goo.com` and `project: github.com`
i wonder if this is too aggressive? 

also, it looks like you found a bug!
type `goo.co` character by character. notice that it doesn't throw an error. that's because `validate`  is referencing a stale value (onChange only dispatches a change to update the value, the value isn't updated until the next cycle).

suggestions:
- L112 in useForm should return `{error: undefined, value}`
- validateOrgName & validateTeamName should use the supplied value instead of grabbing the value from the stale state (e.g. in L77)

**Comment by nickoferrall** [comment]
> i wonder if this is too aggressive?

I think we should throw an error for both `foo.goo.com` and `project: github.com` as they both appear as hyperlinks in invitation emails which could be a security risk.

I've implemented your suggestions and I've also removed the `dirty` check so that we show the error message as the user types rather than `onBlur`.

**Comment by mattkrick** [comment]
I'm cool with that! LGTM

**Pull request merged**